### PR TITLE
enable using arbitrary containers for rbe_autoconfig

### PR DESCRIPTION
**Pull request by nlopezgi**

### CODE REVIEW

File: rules/BUILD.platform.tpl [hunk 1]
```diff
@@ -50,13 +50,8 @@ toolchain(
...
-alias(
-    name = "rbe_ubuntu1604",
-    actual = ":rbe_ubuntu1604_%{revision}",
-)
-
 platform(
-    name = "rbe_ubuntu1604_%{revision}",
+    name = "rbe_platform",
     constraint_values = [
         "@bazel_tools//platforms:x86_64",
```
Comment (xingao267) [inline rules/BUILD.platform.tpl]: The constraints might not properly describe the arbitrary container though.

File: rules/rbe_repo.bzl [hunk 2]
```diff
@@ -12,18 +12,25 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""Repository Rules to generate toolchain configs for an RBE-ubuntu16-04 container.
+"""Repository Rules to generate toolchain configs for a container image.
+
+The toolchain configs (+platform) produced by this rule can be used to, e.g.,
```
Comment (xingao267) [inline rules/rbe_repo.bzl]: (+ platform)

File: rules/BUILD.platform.tpl [hunk 3]
```diff
@@ -50,13 +50,8 @@ toolchain(
...
 )
 
-alias(
-    name = "rbe_ubuntu1604",
-    actual = ":rbe_ubuntu1604_%{revision}",
-)
-
 platform(
-    name = "rbe_ubuntu1604_%{revision}",
+    name = "rbe_platform",
```
Comment (xingao267) [inline rules/BUILD.platform.tpl]: why not just name it as `platform`

File: rules/rbe_repo.bzl [hunk 4]
```diff
@@ -188,9 +217,10 @@ def _impl(ctx):
     name = ctx.attr.name
     outputs_tar = ctx.attr.name + "_out.tar"
 
+    image_id = ctx.attr.registry + "/" + ctx.attr.repository + "@" + ctx.attr.digest
```
Comment (xingao267) [inline rules/rbe_repo.bzl]: image_id is confusing. Maybe image_url

File: rules/rbe_repo.bzl [hunk 5]
```diff
@@ -456,8 +502,7 @@ _rbe_autoconfig = repository_rule(
         ),
         "digest": attr.string(
             mandatory = True,
```
Comment (xingao267) [inline rules/rbe_repo.bzl]: Add a example to make it explict that `sha256` needs to be included in the string.

File: rules/rbe_repo.bzl [hunk 6]
```diff
@@ -488,9 +533,18 @@ _rbe_autoconfig = repository_rule(
...
+            default = _RBE_UBUNTU_REGISTRY,
+            doc = ("The registry to pull the container from. Default is " +
+                   "value for rbe-ubuntu16-04 image."),
+        ),
+        "repository": attr.string(
+            default = _RBE_UBUNTU_REPO,
+            doc = ("The repository to pull the container from. Default is " +
+                   " value for the rbe-ubuntu16-04 image."),
+        ),
         "revision": attr.string(
```
Comment (xingao267) [inline rules/rbe_repo.bzl]: add an example as well.

File: rules/rbe_repo.bzl [hunk 7]
```diff
@@ -523,22 +580,44 @@ def rbe_autoconfig(
...
           Use only if output_base is defined.
-      revision: a revision of an rbe-ubuntu16-04 container.
+      registry: Optional. The registry from which to pull the base image.
+          Should only be set if a custom container is required.
+          Must be set together with digest and repository.
+      repository: Optional. he `repository` of images to pull from.
+          Should only be set if a custom container is required.
+          Must be set together with registry and digest.
+      revision: Optional. A revision of an rbe-ubuntu16-04 container to use.
+          Should not be set if repository, registry and digest are used.
```
Comment (xingao267) [inline rules/rbe_repo.bzl]: add check for this

File: rules/rbe_repo.bzl (hunk 4 above)
Comment (nlopezgi) [reply rules/rbe_repo.bzl]: replaced for image_id

File: rules/rbe_repo.bzl (hunk 2 above)
Comment (nlopezgi) [reply rules/rbe_repo.bzl]: done

**Comment by nlopezgi** [comment]
addressed all review comments. ptal.

File: rules/rbe_repo.bzl [hunk 8]
```diff
@@ -523,22 +622,53 @@ def rbe_autoconfig(
...
+          running the Bazel command to generate the toolchain configs.
+          Set to values for rbe-ubuntu16-04 container.
+          Does not need to be set if your custom container extends
+          an rbe-ubuntu16-04 container.
+          Should be overriden if a custom container does not extend
+          rbe-ubuntu16-04.
     """
     if output_base == "" and config_dir != "":
         fail("config_dir can only be used when output_base is set.")
+    if revision != "latest" and (digest or repository or registry):
```
Comment (xingao267) [inline rules/rbe_repo.bzl]: revision != ""? or not revision?

File: rules/rbe_repo.bzl [hunk 9]
```diff
@@ -523,22 +622,53 @@ def rbe_autoconfig(
...
+          target_compatible_with attr.
+      env: dict. Optional. Additional env variables that will be set when
+          running the Bazel command to generate the toolchain configs.
+          Set to values for rbe-ubuntu16-04 container.
+          Does not need to be set if your custom container extends
+          an rbe-ubuntu16-04 container.
+          Should be overriden if a custom container does not extend
+          rbe-ubuntu16-04.
     """
     if output_base == "" and config_dir != "":
```
Comment (xingao267) [inline rules/rbe_repo.bzl]: actually, can we use not config_dir for config_dir != "" and same for others?

File: rules/rbe_repo.bzl [hunk 10]
```diff
@@ -170,7 +208,7 @@ def _impl(ctx):
     project_root = ctx.os.environ.get(_RBE_AUTOCONF_ROOT, None)
     use_default_project = False
     if not project_root:
-        if ctx.attr.output_base != "":
+        if ctx.attr.output_base:
             fail(("%s env variable must be set for rbe_autoconfig" +
                   " to function properly when output_base is set") % _RBE_AUTOCONF_ROOT)
```
Comment (xingao267) [inline rules/rbe_repo.bzl]: At line 221, %s is not used.

**Pull request merged**